From 3ddd63a1f6c4d281ba9d65b4e74a9a0be3060495 Mon Sep 17 00:00:00 2001
From: Hongcheng Xie <hongcheng.xie@intel.com>
Date: Fri, 26 Mar 2021 10:42:08 +0800
Subject: [PATCH] Move home activity to top with correct user id

When enabled concurent multi user feature, in order to dynamically
start or stop non-system user, we still use system user as framework
current user even in headless system user mode, however that will
cause framework move u0 home activity to top when no others activity
show in the forground.

Make sure get correct user id when get home activity and move to top.

Change-Id: I1b1df27a47517626c6e972710c26798cb57f9aac
Tracked-On: ACP-1371
Signed-off-by: Hongcheng Xie <hongcheng.xie@intel.com>
---
 .../android/server/wm/TaskDisplayArea.java    | 27 ++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/wm/TaskDisplayArea.java b/services/core/java/com/android/server/wm/TaskDisplayArea.java
index 79f3b8340b2..e798f851917 100644
--- a/services/core/java/com/android/server/wm/TaskDisplayArea.java
+++ b/services/core/java/com/android/server/wm/TaskDisplayArea.java
@@ -51,16 +51,23 @@ import android.app.WindowConfiguration;
 import android.content.Intent;
 import android.content.pm.ActivityInfo;
 import android.content.pm.ApplicationInfo;
+import android.os.RemoteException;
+import android.os.ServiceManager;
 import android.os.UserHandle;
+import android.os.UserManager;
+import android.os.UserManagerInternal;
 import android.util.IntArray;
 import android.util.Slog;
 import android.view.SurfaceControl;
 import android.window.WindowContainerTransaction;
 
 import com.android.internal.annotations.VisibleForTesting;
+import com.android.internal.os.IConcurrentMultiUser;
+import com.android.internal.os.RoSystemProperties;
 import com.android.internal.util.ToBooleanFunction;
 import com.android.internal.util.function.pooled.PooledLambda;
 import com.android.internal.util.function.pooled.PooledPredicate;
+import com.android.server.LocalServices;
 import com.android.server.protolog.common.ProtoLog;
 
 import java.io.PrintWriter;
@@ -1613,7 +1620,25 @@ final class TaskDisplayArea extends DisplayArea<ActivityStack> {
 
     @Nullable
     ActivityRecord getHomeActivity() {
-        return getHomeActivityForUser(mRootWindowContainer.mCurrentUser);
+        int userId = mRootWindowContainer.mCurrentUser;
+        if (RoSystemProperties.SUPPORT_CONCURRENT_USER
+                && UserManager.isHeadlessSystemUserMode()) {
+            int id = -1;
+            IConcurrentMultiUser cmu = IConcurrentMultiUser.Stub.asInterface(
+                    ServiceManager.getService("concurrent_user_service"));
+            if (cmu != null) {
+                try {
+                    id = cmu.getUserIdForDisplayId(getDisplayId());
+                } catch (RemoteException e) {
+                    Slog.e(TAG_WM, "getHomeActivity() exception when get user " + id);
+                }
+            }
+            final UserManagerInternal um = LocalServices.getService(UserManagerInternal.class);
+            if (um.isUserRunning(id)) {
+                userId = id;
+            }
+        }
+        return getHomeActivityForUser(userId);
     }
 
     @Nullable
-- 
2.17.1

